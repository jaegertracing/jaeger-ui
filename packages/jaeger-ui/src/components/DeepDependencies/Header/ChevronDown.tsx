// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { IoChevronDown } from 'react-icons/io5';

import './ChevronDown.css';

type TProps = {
  className?: string;
  style?: React.CSSProperties;
};

function ChevronDown(props: TProps) {
  const { className, style } = props;
  const cls = `Ddg--Header--ChevronDown ${className || ''}`;
  return <IoChevronDown className={cls} style={style} />;
}

export default React.memo(ChevronDown);
