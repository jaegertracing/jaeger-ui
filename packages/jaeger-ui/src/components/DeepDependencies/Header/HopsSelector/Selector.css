/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.HopsSelector--Selector {
  color: var(--tx-color-muted);
  margin: 0 0.5em;
  padding: 0px 3px;
}

.HopsSelector--Selector--decrement,
.HopsSelector--Selector--increment {
  background-color: #f4f4f4;
  border: none;
  cursor: pointer;
  height: 27px;
  padding: 0em 0.3em 0.3em;
  width: 27px;
}

.HopsSelector--Selector--decrement {
  margin-right: 0.75em;
}

.HopsSelector--Selector--increment {
  margin-left: 0.75em;
}

.HopsSelector--Selector--decrement:disabled,
.HopsSelector--Selector--increment:disabled {
  color: #bbb;
}

.HopsSelector--Selector--btn {
  background: transparent;
  border: none;
  color: #096dd9;
  cursor: pointer;
  font-weight: bold;
  margin: 0 0.2em;
  padding: 1px 2px;
}

.HopsSelector--Selector--btn.is-Empty {
  color: #aaa;
  font-weight: 400;
}

.HopsSelector--Selector--btn.is-Partial {
  color: #d46b08;
}

/* .HopsSelector--Selector > .HopsSelector--Selector--furthest {
  margin-left: 0.3em;
} */

.HopsSelector--Selector > .HopsSelector--Selector--delimiter {
  color: #999;
  font-weight: 400;
}

.HopsSelector--Selector--slash {
  color: #999;
}

.HopsSelector--Selector--AscIcon {
  margin-right: 0.3em;
}

.HopsSelector--Selector--AscIcon.is-Up {
  transform: scaleY(-1);
}

.HopsSelector--Selector--ChevronDown {
  color: #999;
  height: 0.7em;
  padding-bottom: 0.15em;
  width: 0.7em;
}

.HopsSelector--Selector--ChevronRight {
  color: #999;
  height: 0.7em;
  padding-bottom: 0.15em;
  width: 0.7em;
}

.HopsSelector--Selector--ChevronRight.is-Empty {
  color: #ccc;
}
