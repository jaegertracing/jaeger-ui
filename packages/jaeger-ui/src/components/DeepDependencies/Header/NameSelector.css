/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.NameSelector {
  border-bottom: 1px solid transparent;
  color: var(--tx-color-muted);
  cursor: pointer;
  font-size: 1.5em;
  margin: 0 1.3em 0 0;
  outline: 4px solid transparent;
}

.NameSelector.is-invalid {
  background: #fff2e888;
  border-color: #ffbb96;
  outline-color: #fff2e888;
}

.NameSelector.is-invalid:hover {
  background: #fff2e8;
  border-color: #ff9c6e;
  outline-color: #fff2e8;
}

.NameSelector:hover {
  background: #f4f4f4;
  border-color: #ccc;
  color: var(--tx-color-body);
  outline-color: #f4f4f4;
}

.NameSelector.is-active {
  background: #ddd;
  border-color: #bbb;
  color: var(--tx-color-body);
  outline-color: #ddd;
}

.NameSelector--label {
  font-weight: 400;
  padding-right: 0.5em;
}

.NameSelector--value {
  color: var(--tx-color-title);
}

.NameSelector--chevron {
  font-size: 0.6em;
  margin-left: 0.8em;
  position: relative;
}

.NameSelector--overlay .ant-popover-arrow {
  border: 1px solid #999;
  z-index: -1;
  background: #bbb;
}

.NameSelector--overlay .ant-popover-inner {
  border: 1px solid #ccc;
  overflow: hidden;
}
