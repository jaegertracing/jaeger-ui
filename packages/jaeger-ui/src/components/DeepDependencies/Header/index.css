/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.DdgHeader {
  align-items: stretch;
  background: #fafafa;
  border-bottom: 1px solid #ddd;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1);
  display: flex;
  flex-direction: column;
  position: relative;
  z-index: 10;
}

.DdgHeader--controlHeader {
  align-items: center;
  background-color: #f6f6f6;
  display: flex;
  justify-content: space-between;
  padding-left: 1.25rem;
}

.DdgHeader--paramsHeader {
  align-items: center;
  background: #fafafa;
  border-bottom: 1px solid #eee;
  display: flex;
  flex: 1;
  flex-wrap: wrap;
  font-size: unset;
  margin: 0;
  padding: 0.75rem 1.25rem 0.75rem 1.25rem;
  position: relative;
}

.DdgHeader--findWrapper {
  display: flex;
  flex: 1 1 auto;
  justify-content: flex-end;
}

.DdgHeader--uiFind {
  align-items: center;
  background-color: #fff;
  border-left: 1px solid #eee;
  border-right: 1px solid #eee;
  color: #bbb;
  display: flex;
  flex-grow: 1;
  max-width: 400px;
  min-width: 200px;
  transition: max-width 0.5s;
}

.DdgHeader--uiFind:focus-within {
  max-width: 550px;
}

.DdgHeader--uiFindSearchIcon {
  margin: 0 0.25em;
  padding: 0.24em;
}

.DdgHeader--uiFindInput > input:focus,
.DdgHeader--uiFindInput > input {
  border: none;
  box-shadow: none;
  border-radius: 0px;
}

.DdgHeader--uiFindInfo {
  background: #ffffb8;
  border: 1px dashed #fadb14;
  border-radius: 4px;
  color: #ad8b00;
  margin-left: 0.25em;
  margin-right: 0.5em;
  padding: 0.27em 0.4em;
  stroke: #666;
  stroke-width: 1.2px;
  text-align: center;
  transition: width 0.5s;
  white-space: nowrap;
}

.DdgHeader--uiFindInfo--hidden {
  border-left: dashed var(--tx-color-muted) 1px;
  margin-left: 0.3em;
  padding-left: 0.3em;
}

.DdgHeader--uiFindInfo--icon {
  margin-bottom: 1.5px;
  margin-left: 0.2em;
}

.DdgHeader--uiFindInfo--tooltip {
  white-space: nowrap;
}
