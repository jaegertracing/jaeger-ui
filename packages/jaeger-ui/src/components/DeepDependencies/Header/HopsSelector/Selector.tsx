// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';
import { Popover } from 'antd';
import { ImSortAmountAsc } from 'react-icons/im';
import { FaChevronRight } from 'react-icons/fa';

import ChevronDown from '../ChevronDown';
import { trackHopChange } from '../../index.track';
import { ECheckedStatus, EDirection, THop } from '../../../../model/ddg/types';

import './Selector.css';

type TProps = {
  direction: EDirection;
  furthestDistance: number;
  furthestFullDistance: number;
  furthestFullness: ECheckedStatus;
  handleClick: (distance: number, direction: EDirection) => void;
  hops: THop[];
};

const CLASSNAME = 'HopsSelector--Selector';

export default class Selector extends PureComponent<TProps> {
  private handleClick(distance: number) {
    const { direction, furthestFullDistance, handleClick } = this.props;
    handleClick(distance, direction);
    trackHopChange(furthestFullDistance, distance, direction);
  }

  private makeBtn = (
    { distance, fullness, suffix = 'popover-content' }: THop & { suffix?: string },
    showChevron?: number
  ) => {
    const { direction } = this.props;
    return (
      <React.Fragment key={`${distance} ${direction} ${suffix}`}>
        {Boolean(showChevron) && <FaChevronRight className={`${CLASSNAME}--ChevronRight is-${fullness}`} />}
        <button
          className={`${CLASSNAME}--btn is-${fullness} ${CLASSNAME}--${suffix}`}
          type="button"
          onClick={() => this.handleClick(distance)}
        >
          {Math.abs(distance)}
        </button>
      </React.Fragment>
    );
  };

  render() {
    const { direction, furthestDistance, furthestFullness, handleClick, hops } = this.props;
    const streamText = direction === EDirection.Downstream ? 'Down' : 'Up';
    const streamLabel = `${streamText}stream hops`;
    const lowercaseLabel = streamLabel.toLowerCase();
    if (hops.length <= 1) {
      return <span className={CLASSNAME}>No {lowercaseLabel}</span>;
    }

    const decrementBtn = (
      <button
        key={`decrement ${direction}`}
        disabled={furthestDistance === 0}
        className={`${CLASSNAME}--decrement`}
        type="button"
        onClick={() => handleClick(furthestDistance - direction, direction)}
      >
        -
      </button>
    );
    const incrementBtn = (
      <button
        key={`increment ${direction}`}
        disabled={furthestDistance === hops[hops.length - 1].distance}
        className={`${CLASSNAME}--increment`}
        type="button"
        onClick={() => handleClick(furthestDistance + direction, direction)}
      >
        +
      </button>
    );
    const delimiterBtn = this.makeBtn({
      ...hops[hops.length - 1],
      suffix: 'delimiter',
    });
    const furthestBtn = this.makeBtn({
      distance: furthestDistance,
      fullness: furthestFullness,
      suffix: 'furthest',
    });

    return (
      <Popover
        arrowPointAtCenter
        content={[decrementBtn, ...hops.map(this.makeBtn), incrementBtn]}
        placement="bottom"
        title={`Visible ${lowercaseLabel}`}
      >
        <span className={CLASSNAME}>
          <ImSortAmountAsc className={`${CLASSNAME}--AscIcon is-${streamText}`} />
          {streamLabel}
          {furthestBtn}
          <span className={`${CLASSNAME}--delimiter`}>/</span>
          {delimiterBtn}
          {/* <IoChevronDown className={`${CLASSNAME}--ChevronDown`} /> */}
          <ChevronDown className="ub-ml1" />
        </span>
      </Popover>
    );
  }
}
