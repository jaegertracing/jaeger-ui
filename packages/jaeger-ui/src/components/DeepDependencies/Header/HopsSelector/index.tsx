// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { memo } from 'react';

import { decode } from '../../../../model/ddg/visibility-codec';
import { ECheckedStatus, EDirection, TDdgDistanceToPathElems, THop } from '../../../../model/ddg/types';
import Selector from './Selector';

type TProps = {
  distanceToPathElems?: TDdgDistanceToPathElems;
  handleClick: (distance: number, direction: EDirection) => void;
  visEncoding?: string;
};

export default memo(function HopsSelector({ distanceToPathElems, handleClick, visEncoding }: TProps) {
  if (!distanceToPathElems) return <div />;

  const downstreamHops: THop[] = [];
  const upstreamHops: THop[] = [];
  let minFullDistance = 0;
  let minVisDistance = 0;
  let minVisDistanceFullness = ECheckedStatus.Empty;
  let maxFullDistance = 0;
  let maxVisDistance = 0;
  let maxVisDistanceFullness = ECheckedStatus.Empty;
  const visibleIndices = visEncoding && new Set(decode(visEncoding));

  distanceToPathElems.forEach((elems, distance) => {
    let fullness: ECheckedStatus;
    if (visibleIndices) {
      const visible = elems.filter(({ visibilityIdx }) => visibleIndices.has(visibilityIdx));
      if (visible.length === elems.length) {
        fullness = ECheckedStatus.Full;
      } else if (!visible.length) {
        fullness = ECheckedStatus.Empty;
      } else {
        fullness = ECheckedStatus.Partial;
      }
    } else {
      fullness = Math.abs(distance) <= 2 ? ECheckedStatus.Full : ECheckedStatus.Empty;
    }

    if (distance >= 0) downstreamHops.push({ distance, fullness });
    if (distance <= 0) upstreamHops.push({ distance, fullness });

    if (fullness !== ECheckedStatus.Empty) {
      if (distance >= maxVisDistance) {
        maxVisDistance = distance;
        maxVisDistanceFullness = fullness;
      }
      if (distance <= minVisDistance) {
        minVisDistance = distance;
        minVisDistanceFullness = fullness;
      }
    }

    if (fullness === ECheckedStatus.Full) {
      if (distance >= maxFullDistance) maxFullDistance = distance;
      if (distance <= minFullDistance) minFullDistance = distance;
    }
  });

  downstreamHops.sort(({ distance: a }, { distance: b }) => a - b);
  upstreamHops.sort(({ distance: a }, { distance: b }) => b - a);

  return (
    <div>
      <Selector
        direction={EDirection.Upstream}
        handleClick={handleClick}
        hops={upstreamHops}
        furthestDistance={minVisDistance}
        furthestFullDistance={minFullDistance}
        furthestFullness={minVisDistanceFullness}
      />
      <Selector
        direction={EDirection.Downstream}
        handleClick={handleClick}
        hops={downstreamHops}
        furthestDistance={maxVisDistance}
        furthestFullDistance={maxFullDistance}
        furthestFullness={maxVisDistanceFullness}
      />
    </div>
  );
});
