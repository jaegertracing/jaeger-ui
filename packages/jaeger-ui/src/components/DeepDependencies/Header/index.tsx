// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Icon, Input, Tooltip } from 'antd';
import MdVisibility from 'react-icons/lib/md/visibility';
import MdVisibilityOff from 'react-icons/lib/md/visibility-off';

import HopsSelector from './HopsSelector';
import NameSelector from '../../common/NameSelector';
import LayoutSettings from './LayoutSettings';
import { trackFilter, trackHeaderSetOperation, trackShowMatches } from '../index.track';
import UiFindInput from '../../common/UiFindInput';
import { EDirection, TDdgDistanceToPathElems, EDdgDensity } from '../../../model/ddg/types';

import './index.css';

type TProps = {
  clearOperation: () => void;
  density: EDdgDensity;
  distanceToPathElems?: TDdgDistanceToPathElems;
  hiddenUiFindMatches?: Set<string>;
  operation?: string;
  operations: string[] | undefined;
  service?: string;
  services?: string[] | null;
  setDensity: (density: EDdgDensity) => void;
  setDistance: (distance: number, direction: EDirection) => void;
  setOperation: (operation: string) => void;
  setService: (service: string) => void;
  showOperations: boolean;
  showParameters?: boolean;
  showVertices: (vertexKeys: string[]) => void;
  toggleShowOperations: (enable: boolean) => void;
  uiFindCount: number | undefined;
  visEncoding?: string;
};
export default class Header extends React.PureComponent<TProps> {
  private _uiFindInput: React.RefObject<Input> = React.createRef();

  static defaultProps = {
    showParameters: true,
  };

  focusUiFindInput = () => {
    if (this._uiFindInput.current) {
      this._uiFindInput.current.focus();
    }
  };

  getUiFindInfo = () => {
    const { hiddenUiFindMatches, uiFindCount } = this.props;

    if (uiFindCount === undefined) return null;

    let hasHidden = false;
    let hiddenInfo: React.ReactNode = null;
    let tipText = uiFindCount ? 'All matches are visible' : 'No matches';
    if (hiddenUiFindMatches && hiddenUiFindMatches.size) {
      const { size } = hiddenUiFindMatches;
      hasHidden = true;
      tipText = `Click to view ${size} hidden match${size !== 1 ? 'es' : ''}`;
      hiddenInfo = (
        <span className="DdgHeader--uiFindInfo--hidden">
          {size}
          <MdVisibilityOff className="DdgHeader--uiFindInfo--icon" />
        </span>
      );
    }

    return (
      <Tooltip overlayClassName="DdgHeader--uiFindInfo--tooltip" placement="topRight" title={tipText}>
        {/* arbitrary span is necessary as Tooltip alters child's styling */}
        <span>
          <button
            className="DdgHeader--uiFindInfo"
            disabled={!hasHidden}
            onClick={this.handleInfoClick}
            type="button"
          >
            {uiFindCount}
            {(uiFindCount !== 0 || hasHidden) && <MdVisibility className="DdgHeader--uiFindInfo--icon" />}
            {hiddenInfo}
          </button>
        </span>
      </Tooltip>
    );
  };

  setOperation = (operation: string) => {
    trackHeaderSetOperation();
    this.props.setOperation(operation);
  };

  handleInfoClick = () => {
    trackShowMatches();
    const { hiddenUiFindMatches, showVertices } = this.props;
    if (hiddenUiFindMatches) showVertices(Array.from(hiddenUiFindMatches));
  };

  render() {
    const {
      clearOperation,
      density,
      distanceToPathElems,
      operation,
      operations,
      service,
      services,
      setDensity,
      setDistance,
      setService,
      showOperations,
      showParameters,
      toggleShowOperations,
      visEncoding,
    } = this.props;

    return (
      <header className="DdgHeader">
        {showParameters && (
          <div className="DdgHeader--paramsHeader">
            <NameSelector
              label="Service"
              placeholder="Select a service…"
              value={service || null}
              setValue={setService}
              required
              options={services || []}
            />
            {service && (
              <NameSelector
                clearValue={clearOperation}
                label="Operation"
                placeholder="Filter by operation…"
                value={operation || null}
                setValue={this.setOperation}
                options={operations || []}
              />
            )}
          </div>
        )}
        <div className="DdgHeader--controlHeader">
          <LayoutSettings
            density={density}
            setDensity={setDensity}
            showOperations={showOperations}
            toggleShowOperations={toggleShowOperations}
          />
          <HopsSelector
            distanceToPathElems={distanceToPathElems}
            handleClick={setDistance}
            visEncoding={visEncoding}
          />
          <div className="DdgHeader--findWrapper">
            <div className="DdgHeader--uiFind" role="button" onClick={this.focusUiFindInput}>
              <Icon className="DdgHeader--uiFindSearchIcon" type="search" />
              <UiFindInput
                allowClear
                forwardedRef={this._uiFindInput}
                inputProps={{ className: 'DdgHeader--uiFindInput' }}
                trackFindFunction={trackFilter}
              />
              {this.getUiFindInfo()}
            </div>
          </div>
        </div>
      </header>
    );
  }
}
