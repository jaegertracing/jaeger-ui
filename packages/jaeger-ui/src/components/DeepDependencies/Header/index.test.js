// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';
import { Tooltip } from 'antd';
import MdVisibility from 'react-icons/lib/md/visibility';
import MdVisibilityOff from 'react-icons/lib/md/visibility-off';

import Header from './index';
import HopsSelector from './HopsSelector';
import NameSelector from '../../common/NameSelector';
import * as track from '../index.track';

describe('<Header>', () => {
  const minProps = {
    clearOperation: () => {},
    setDistance: () => {},
    setOperation: jest.fn(),
    setService: () => {},
  };
  const service = 'testService';
  const services = [service];
  const operation = 'testOperation';
  const operations = [operation];
  let trackSetOpSpy;
  let wrapper;

  beforeAll(() => {
    trackSetOpSpy = jest.spyOn(track, 'trackHeaderSetOperation');
  });

  beforeEach(() => {
    wrapper = shallow(<Header {...minProps} />);
  });

  it('renders with minimal props', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('omits the operation selector if a service is not selected', () => {
    const nameSelector = wrapper.find(NameSelector);
    expect(nameSelector.length).toBe(1);
    expect(nameSelector.prop('label')).toMatch(/service/i);
  });

  it('renders the operation selector iff a service is selected', () => {
    let nameSelector = wrapper.find(NameSelector);
    expect(nameSelector.length).toBe(1);
    wrapper.setProps({ service, services });
    nameSelector = wrapper.find(NameSelector);
    expect(nameSelector.length).toBe(2);
    expect(nameSelector.at(1).prop('label')).toMatch(/operation/i);
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ operation, operations });
    expect(wrapper).toMatchSnapshot();
  });

  it('tracks when operation selector sets a value', () => {
    wrapper.setProps({ service, services });
    const testOp = 'test operation';
    expect(trackSetOpSpy).not.toHaveBeenCalled();

    wrapper
      .find(NameSelector)
      .at(1)
      .prop('setValue')(testOp);
    expect(trackSetOpSpy).toHaveBeenCalledTimes(1);
    expect(minProps.setOperation).toHaveBeenCalledWith(testOp);
  });

  it('renders the hops selector if distanceToPathElems is provided', () => {
    wrapper.setProps({
      distanceToPathElems: new Map(),
      visEncoding: '3',
    });
    expect(wrapper.find(HopsSelector).length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('focuses uiFindInput IFF rendered when clicking on wrapping div', () => {
    const click = () => wrapper.find('.DdgHeader--uiFind').simulate('click');
    const focus = jest.fn();
    click();

    wrapper.instance()._uiFindInput = { current: { focus } };
    click();
    expect(focus).toHaveBeenCalledTimes(1);
  });

  describe('uiFind match information', () => {
    const getBtn = () => wrapper.find('button');
    const getMatchesInfo = () => wrapper.find('.DdgHeader--uiFindInfo');
    const getTooltip = () => wrapper.find(Tooltip);
    const hiddenUiFindMatches = new Set(['hidden', 'match', 'vertices']);
    const uiFindCount = 20;
    const expectedFindCount = expect.stringContaining(`${uiFindCount}`);
    const expectedHiddenCount = expect.stringContaining(`${hiddenUiFindMatches.size}`);
    const expectedHiddenTitle = `Click to view ${hiddenUiFindMatches.size} hidden matches`;

    it('renders no info if count is `undefined`', () => {
      expect(getMatchesInfo()).toHaveLength(0);
      expect(getTooltip()).toHaveLength(0);
    });

    it('renders count if `hiddenUiFindMatches` is `undefined` or empty', () => {
      const expectedTitle = 'All matches are visible';

      wrapper.setProps({ uiFindCount });
      expect(getMatchesInfo().text()).toEqual(expectedFindCount);
      expect(getMatchesInfo().text()).not.toEqual(expectedHiddenCount);
      expect(getTooltip().prop('title')).toBe(expectedTitle);
      expect(getBtn().prop('disabled')).toBe(true);
      expect(wrapper.find(MdVisibility)).toHaveLength(1);
      expect(wrapper.find(MdVisibilityOff)).toHaveLength(0);

      wrapper.setProps({ hiddenUiFindMatches: new Set() });
      expect(getMatchesInfo().text()).toEqual(expectedFindCount);
      expect(getMatchesInfo().text()).not.toEqual(expectedHiddenCount);
      expect(getTooltip().prop('title')).toBe(expectedTitle);
      expect(getBtn().prop('disabled')).toBe(true);
      expect(wrapper.find(MdVisibility)).toHaveLength(1);
      expect(wrapper.find(MdVisibilityOff)).toHaveLength(0);
    });

    it('renders both visible and hidden counts if both are provided', () => {
      wrapper.setProps({ hiddenUiFindMatches, uiFindCount });

      expect(getMatchesInfo().text()).toEqual(expectedFindCount);
      expect(getMatchesInfo().text()).toEqual(expectedHiddenCount);
      expect(getTooltip().prop('title')).toBe(expectedHiddenTitle);
      expect(getBtn().prop('disabled')).toBe(false);
      expect(wrapper.find(MdVisibility)).toHaveLength(1);
      expect(wrapper.find(MdVisibilityOff)).toHaveLength(1);
    });

    it('renders 0 with correct tooltip if there are no visible nor hidden matches', () => {
      const expectedTitle = 'No matches';
      wrapper.setProps({ uiFindCount: 0 });

      expect(getMatchesInfo().text()).toBe('0');
      expect(getTooltip().prop('title')).toBe(expectedTitle);
      expect(getBtn().prop('disabled')).toBe(true);
      expect(wrapper.find(MdVisibility)).toHaveLength(0);
      expect(wrapper.find(MdVisibilityOff)).toHaveLength(0);
    });

    it('renders 0 with correct tooltip if there are no matches but there are hidden matches', () => {
      wrapper.setProps({ uiFindCount: 0 });
      wrapper.setProps({ hiddenUiFindMatches });

      expect(getMatchesInfo().text()).toEqual(expect.stringContaining('0'));
      expect(getMatchesInfo().text()).toEqual(expectedHiddenCount);
      expect(getTooltip().prop('title')).toBe(expectedHiddenTitle);
      expect(getBtn().prop('disabled')).toBe(false);
      expect(wrapper.find(MdVisibility)).toHaveLength(1);
      expect(wrapper.find(MdVisibilityOff)).toHaveLength(1);
    });

    it('renders correct plurality in tooltip', () => {
      const expectedTitle = 'Click to view 1 hidden match';
      wrapper.setProps({ hiddenUiFindMatches: new Set([Array.from(hiddenUiFindMatches)[0]]), uiFindCount });

      expect(getTooltip().prop('title')).toBe(expectedTitle);
    });

    it('calls props.showVertices with vertices in props.hiddenUiFindMatches when clicked with hiddenUiFindMatches', () => {
      const showVertices = jest.fn();
      wrapper.setProps({ showVertices, uiFindCount });
      getBtn().simulate('click');
      expect(showVertices).toHaveBeenCalledTimes(0);

      wrapper.setProps({ hiddenUiFindMatches });
      getBtn().simulate('click');
      expect(showVertices).toHaveBeenCalledTimes(1);
      expect(showVertices).toHaveBeenCalledWith(Array.from(hiddenUiFindMatches));
    });
  });
});
