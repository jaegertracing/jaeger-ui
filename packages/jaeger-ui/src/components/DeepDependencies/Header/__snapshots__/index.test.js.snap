// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header> renders the hops selector if distanceToPathElems is provided 1`] = `
<header
  className="DdgHeader"
>
  <div
    className="DdgHeader--paramsHeader"
  >
    <NameSelector
      label="Service"
      options={Array []}
      placeholder="Select a service…"
      required={true}
      setValue={[Function]}
      value={null}
    />
  </div>
  <div
    className="DdgHeader--controlHeader"
  >
    <LayoutSettings />
    <Memo(HopsSelector)
      distanceToPathElems={Map {}}
      handleClick={[Function]}
      visEncoding="3"
    />
    <div
      className="DdgHeader--findWrapper"
    >
      <div
        className="DdgHeader--uiFind"
        onClick={[Function]}
        role="button"
      >
        <ForwardRef(SearchOutlined)
          className="DdgHeader--uiFindSearchIcon"
        />
        <withRouter(Connect(UnconnectedUiFindInput))
          allowClear={true}
          forwardedRef={
            Object {
              "current": null,
            }
          }
          inputProps={
            Object {
              "className": "DdgHeader--uiFindInput",
            }
          }
          trackFindFunction={[Function]}
        />
      </div>
    </div>
  </div>
</header>
`;

exports[`<Header> renders the operation selector iff a service is selected 1`] = `
<header
  className="DdgHeader"
>
  <div
    className="DdgHeader--paramsHeader"
  >
    <NameSelector
      label="Service"
      options={
        Array [
          "testService",
        ]
      }
      placeholder="Select a service…"
      required={true}
      setValue={[Function]}
      value="testService"
    />
    <NameSelector
      clearValue={[Function]}
      label="Operation"
      options={Array []}
      placeholder="Filter by operation…"
      setValue={[Function]}
      value={null}
    />
  </div>
  <div
    className="DdgHeader--controlHeader"
  >
    <LayoutSettings />
    <Memo(HopsSelector)
      handleClick={[Function]}
    />
    <div
      className="DdgHeader--findWrapper"
    >
      <div
        className="DdgHeader--uiFind"
        onClick={[Function]}
        role="button"
      >
        <ForwardRef(SearchOutlined)
          className="DdgHeader--uiFindSearchIcon"
        />
        <withRouter(Connect(UnconnectedUiFindInput))
          allowClear={true}
          forwardedRef={
            Object {
              "current": null,
            }
          }
          inputProps={
            Object {
              "className": "DdgHeader--uiFindInput",
            }
          }
          trackFindFunction={[Function]}
        />
      </div>
    </div>
  </div>
</header>
`;

exports[`<Header> renders the operation selector iff a service is selected 2`] = `
<header
  className="DdgHeader"
>
  <div
    className="DdgHeader--paramsHeader"
  >
    <NameSelector
      label="Service"
      options={
        Array [
          "testService",
        ]
      }
      placeholder="Select a service…"
      required={true}
      setValue={[Function]}
      value="testService"
    />
    <NameSelector
      clearValue={[Function]}
      label="Operation"
      options={
        Array [
          "testOperation",
        ]
      }
      placeholder="Filter by operation…"
      setValue={[Function]}
      value="testOperation"
    />
  </div>
  <div
    className="DdgHeader--controlHeader"
  >
    <LayoutSettings />
    <Memo(HopsSelector)
      handleClick={[Function]}
    />
    <div
      className="DdgHeader--findWrapper"
    >
      <div
        className="DdgHeader--uiFind"
        onClick={[Function]}
        role="button"
      >
        <ForwardRef(SearchOutlined)
          className="DdgHeader--uiFindSearchIcon"
        />
        <withRouter(Connect(UnconnectedUiFindInput))
          allowClear={true}
          forwardedRef={
            Object {
              "current": null,
            }
          }
          inputProps={
            Object {
              "className": "DdgHeader--uiFindInput",
            }
          }
          trackFindFunction={[Function]}
        />
      </div>
    </div>
  </div>
</header>
`;

exports[`<Header> renders with minimal props 1`] = `
<header
  className="DdgHeader"
>
  <div
    className="DdgHeader--paramsHeader"
  >
    <NameSelector
      label="Service"
      options={Array []}
      placeholder="Select a service…"
      required={true}
      setValue={[Function]}
      value={null}
    />
  </div>
  <div
    className="DdgHeader--controlHeader"
  >
    <LayoutSettings />
    <Memo(HopsSelector)
      handleClick={[Function]}
    />
    <div
      className="DdgHeader--findWrapper"
    >
      <div
        className="DdgHeader--uiFind"
        onClick={[Function]}
        role="button"
      >
        <ForwardRef(SearchOutlined)
          className="DdgHeader--uiFindSearchIcon"
        />
        <withRouter(Connect(UnconnectedUiFindInput))
          allowClear={true}
          forwardedRef={
            Object {
              "current": null,
            }
          }
          inputProps={
            Object {
              "className": "DdgHeader--uiFindInput",
            }
          }
          trackFindFunction={[Function]}
        />
      </div>
    </div>
  </div>
</header>
`;
